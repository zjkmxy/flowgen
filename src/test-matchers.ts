import { execFileSync } from "child_process";

import chalk from "chalk";
import flow from "flow-bin";

import { beautify } from ".";

// eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
    interface Matchers<R> {
      toBeValidFlowTypeDeclarations(): R;
    }
  }
}

expect.extend({
  async toBeValidFlowTypeDeclarations(source) {
    const beautifiedSource = await beautify(source);
    try {
      execFileSync(
        flow,
        ["check-contents", "--color=always", "--timeout=30"],
        {
          input: beautifiedSource,
          stdio: ["pipe", "pipe", "pipe"],
        },
      );
    } catch (err) {
      return {
        message: () =>
          `expected ${chalk.bold(
            beautifiedSource.trimEnd(),
          )} to be valid flow:\n${chalk.red(err.stdout)}`,
        pass: false,
      };
    }

    return {
      message: () =>
        `expected ${chalk.bold(
          beautifiedSource.trimEnd(),
        )} not to be valid flow`,
      pass: true,
    };
  },
});
