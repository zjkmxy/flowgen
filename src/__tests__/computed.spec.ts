import { compiler, beautify } from "..";
import "../test-matchers";

// TODO: fix this test
// The snippets below are neither valid TypeScript nor valid Flow. Shouldn't test on them.

// it("should handle computed Symbol.iterator and Symbol.asyncIterator", async () => {
//   const ts = `
//   type A = {
//     [Symbol.asyncIterator]?(): any,
//     [Symbol.iterator]?(): any,
//     readonly [Symbol.asyncIterator]?(): any,
//     readonly [Symbol.iterator]?(): any,
//     [Symbol.asyncIterator](): any,
//     [Symbol.iterator](): any,
//     readonly [Symbol.asyncIterator](): any,
//     readonly [Symbol.iterator](): any,
//     [Symbol.asyncIterator]?: any,
//     [Symbol.iterator]?: any,
//     readonly [Symbol.asyncIterator]?: any,
//     readonly [Symbol.iterator]?: any,
//     [Symbol.asyncIterator]: any,
//     [Symbol.iterator]: any,
//     readonly [Symbol.asyncIterator]: any,
//     readonly [Symbol.iterator]: any,
//   }
//   declare class B {
//     [Symbol.asyncIterator]?(): any,
//     [Symbol.iterator]?(): any,
//     readonly [Symbol.asyncIterator]?(): any,
//     readonly [Symbol.iterator]?(): any,
//     [Symbol.asyncIterator](): any,
//     [Symbol.iterator](): any,
//     readonly [Symbol.asyncIterator](): any,
//     readonly [Symbol.iterator](): any,
//     [Symbol.asyncIterator]?: any,
//     [Symbol.iterator]?: any,
//     readonly [Symbol.asyncIterator]?: any,
//     readonly [Symbol.iterator]?: any,
//     [Symbol.asyncIterator]: any,
//     [Symbol.iterator]: any,
//     readonly [Symbol.asyncIterator]: any,
//     readonly [Symbol.iterator]: any,
//   }
//   interface C {
//     [Symbol.asyncIterator]?(): any,
//     [Symbol.iterator]?(): any,
//     readonly [Symbol.asyncIterator]?(): any,
//     readonly [Symbol.iterator]?(): any,
//     [Symbol.asyncIterator](): any,
//     [Symbol.iterator](): any,
//     readonly [Symbol.asyncIterator](): any,
//     readonly [Symbol.iterator](): any,
//     [Symbol.asyncIterator]?: any,
//     [Symbol.iterator]?: any,
//     readonly [Symbol.asyncIterator]?: any,
//     readonly [Symbol.iterator]?: any,
//     [Symbol.asyncIterator]: any,
//     [Symbol.iterator]: any,
//     readonly [Symbol.asyncIterator]: any,
//     readonly [Symbol.iterator]: any,
//   }
// `;

//   const result = compiler.compileDefinitionString(ts, { quiet: true });

//   expect(await beautify(result)).toMatchSnapshot();
//   expect(result).toBeValidFlowTypeDeclarations();
// });

// it("should handle string literals", async () => {
//   const ts = `
//   type A = {
//     ["foo"]?(): any,
//     readonly ["foo"]?(): any,
//     ["foo"](): any,
//     readonly ["foo"](): any,
//     ["foo"]?: any,
//     readonly ["foo"]?: any,
//     ["foo"]: any,
//     readonly ["foo"]: any,
//   }
//   declare class B {
//     ["foo"]?(): any,
//     readonly ["foo"]?(): any,
//     ["foo"](): any,
//     readonly ["foo"](): any,
//     ["foo"]?: any,
//     readonly ["foo"]?: any,
//     ["foo"]: any,
//     readonly ["foo"]: any,
//   }
//   interface C {
//     ["foo"]?(): any,
//     readonly ["foo"]?(): any,
//     ["foo"](): any,
//     readonly ["foo"](): any,
//     ["foo"]?: any,
//     readonly ["foo"]?: any,
//     ["foo"]: any,
//     readonly ["foo"]: any,
//   }
// `;

//   const result = compiler.compileDefinitionString(ts, { quiet: true });

//   expect(await beautify(result)).toMatchSnapshot();
//   expect(result).toBeValidFlowTypeDeclarations();
// });

// it("should approximate unsupported keys", async () => {
//   const ts = `
//   type A = {
//     [Foo]?(): any,
//     readonly [Foo]?(): any,
//     [Foo](): any,
//     readonly [Foo](): any,
//     [Foo]?: any,
//     readonly [Foo]?: any,
//     [Foo]: any,
//     readonly [Foo]: any,
//   }
//   declare class B {
//     [Foo]?(): any,
//     readonly [Foo]?(): any,
//     [Foo](): any,
//     readonly [Foo](): any,
//     [Foo]?: any,
//     readonly [Foo]?: any,
//     [Foo]: any,
//     readonly [Foo]: any,
//   }
//   interface C {
//     [Foo]?(): any,
//     readonly [Foo]?(): any,
//     [Foo](): any,
//     readonly [Foo](): any,
//     [Foo]?: any,
//     readonly [Foo]?: any,
//     [Foo]: any,
//     readonly [Foo]: any,
//   }
// `;

//   const result = compiler.compileDefinitionString(ts, { quiet: true });

//   expect(await beautify(result)).toMatchSnapshot();
//   expect(result).not.toBeValidFlowTypeDeclarations(); // unsupported-syntax
// });
