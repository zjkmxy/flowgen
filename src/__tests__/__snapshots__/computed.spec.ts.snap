// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should approximate unsupported keys 1`] = `
"declare type A = {
  [typeof Foo]: (() => any) | void,
  +[typeof Foo]: (() => any) | void,
  [typeof Foo]: () => any,
  +[typeof Foo]: () => any,
  [typeof Foo]: any | void,
  +[typeof Foo]: any | void,
  [typeof Foo]: any,
  +[typeof Foo]: any,
};
declare class B {
  [typeof Foo]: (() => any) | void;
  +[typeof Foo]: (() => any) | void;
  [typeof Foo]: () => any;
  +[typeof Foo]: () => any;
  [typeof Foo]: any | void;
  +[typeof Foo]: any | void;
  [typeof Foo]: any;
  +[typeof Foo]: any;
}
declare interface C {
  [typeof Foo]: (() => any) | void;
  +[typeof Foo]: (() => any) | void;
  [typeof Foo]: () => any;
  +[typeof Foo]: () => any;
  [typeof Foo]: any | void;
  +[typeof Foo]: any | void;
  [typeof Foo]: any;
  +[typeof Foo]: any;
}
"
`;

exports[`should handle computed Symbol.iterator and Symbol.asyncIterator 1`] = `
"declare type A = {
  @@asyncIterator: (() => any) | void,
  @@iterator: (() => any) | void,
  +@@asyncIterator: (() => any) | void,
  +@@iterator: (() => any) | void,
  @@asyncIterator: () => any,
  @@iterator: () => any,
  +@@asyncIterator: () => any,
  +@@iterator: () => any,
  @@asyncIterator: any | void,
  @@iterator: any | void,
  +@@asyncIterator: any | void,
  +@@iterator: any | void,
  @@asyncIterator: any,
  @@iterator: any,
  +@@asyncIterator: any,
  +@@iterator: any,
};
declare class B {
  @@asyncIterator: (() => any) | void;
  @@iterator: (() => any) | void;
  +@@asyncIterator: (() => any) | void;
  +@@iterator: (() => any) | void;
  @@asyncIterator: () => any;
  @@iterator: () => any;
  +@@asyncIterator: () => any;
  +@@iterator: () => any;
  @@asyncIterator: any | void;
  @@iterator: any | void;
  +@@asyncIterator: any | void;
  +@@iterator: any | void;
  @@asyncIterator: any;
  @@iterator: any;
  +@@asyncIterator: any;
  +@@iterator: any;
}
declare interface C {
  @@asyncIterator: (() => any) | void;
  @@iterator: (() => any) | void;
  +@@asyncIterator: (() => any) | void;
  +@@iterator: (() => any) | void;
  @@asyncIterator: () => any;
  @@iterator: () => any;
  +@@asyncIterator: () => any;
  +@@iterator: () => any;
  @@asyncIterator: any | void;
  @@iterator: any | void;
  +@@asyncIterator: any | void;
  +@@iterator: any | void;
  @@asyncIterator: any;
  @@iterator: any;
  +@@asyncIterator: any;
  +@@iterator: any;
}
"
`;

exports[`should handle string literals 1`] = `
"declare type A = {
  foo: (() => any) | void,
  +foo: (() => any) | void,
  foo: () => any,
  +foo: () => any,
  foo: any | void,
  +foo: any | void,
  foo: any,
  +foo: any,
};
declare class B {
  foo: (() => any) | void;
  +foo: (() => any) | void;
  foo: () => any;
  +foo: () => any;
  foo: any | void;
  +foo: any | void;
  foo: any;
  +foo: any;
}
declare interface C {
  foo: (() => any) | void;
  +foo: (() => any) | void;
  foo: () => any;
  +foo: () => any;
  foo: any | void;
  +foo: any | void;
  foo: any;
  +foo: any;
}
"
`;
